src/main/java/org/truffleruby/core/regexp/MatchDataNodes.java

```diff
@@ -425,130 +512,67 @@ public int length(DynamicObject matchData) {
     @CoreMethod(names = "pre_match")
     public abstract static class PreMatchNode extends CoreMethodArrayArgumentsNode {
 
-        @Child private TaintResultNode taintResultNode = new TaintResultNode();
         @Child private RopeNodes.MakeSubstringNode makeSubstringNode = RopeNodes.MakeSubstringNode.create();
+        @Child private AllocateObjectNode allocateNode = AllocateObjectNode.create();
+        @Child private TaintResultNode taintNode = new TaintResultNode();
 
         public static PreMatchNode create() {
             return PreMatchNodeFactory.create(new RubyNode[0]);
         }
 
         public abstract DynamicObject execute(DynamicObject matchData);
 
-        @Specialization(guards = "isConstructed(matchData)")
-        public Object preMatchFast(DynamicObject matchData) {
-            return Layouts.MATCH_DATA.getPre(matchData);
-        }
-
-        @Specialization(guards = "!isConstructed(matchData)")
+        @Specialization
         public Object preMatch(DynamicObject matchData) {
             DynamicObject source = Layouts.MATCH_DATA.getSource(matchData);
             Rope sourceRope = StringOperations.rope(source);
             Region region = Layouts.MATCH_DATA.getRegion(matchData);
             int start = 0;
             int length = region.beg[0];
             Rope rope = makeSubstringNode.executeMake(sourceRope, start, length);
-            DynamicObject newStr = createSubstring(source, rope);
-            Layouts.MATCH_DATA.setPre(matchData, newStr);
-            return taintResultNode.maybeTaint(source, newStr);
-        }
-
-        protected boolean isConstructed(DynamicObject matchData) {
-            return Layouts.MATCH_DATA.getPre(matchData) != null;
+            DynamicObject string = allocateNode.allocate(Layouts.BASIC_OBJECT.getLogicalClass(source), Layouts.STRING.build(false, false, rope));
+            return taintNode.maybeTaint(matchData, string);
         }
     }
 
-    @CoreMethod(names = "post_match")
+    @CoreMethod(names = "post_match", taintFrom = 0)
```

Comment:
Does taintFrom works? If so it seems nicer than doing it manually.
From my 5 seconds testing, Matching against a tainted String makes the MatchData tainted, and also the Strings returned by it, is that right?

Reply:
Curses! `taintFrom` works at the level of core method. So we'd either need to rewrite the match reference nodes for special variables to use `RubyCallNode` or use manual tainting.

Reply:
Ah, because PostMatchNode is used by other nodes like `$'`?
Manual tainting seems the most straightforward then, you just need to read the taint from `self` and give it to  `Layouts.STRING.build()`.

Reply:
Actually changed `ReadMatchReferenceNodes` to use calls where possible, now that they should all be reasonably optimal, since it made the logic there simpler. The only real exception is `$+` which doesn't have quite the same semantics as `$~[-1]`

Reply:
That sounds good too.